### PR TITLE
Notifications overhaul

### LINKED ISSUE #8697: Sync (Push) notifications
Done
- [x] Android
- [x] iOS
- [ ] Desktop
- [x] Server

When receiving a notification:

On Android/iOS:
- you get a push notification (displayed as popup/ in statusbar)
- you can click on it, get redirected to list of notifications
- on iOS you can delete notifications via swipe (get also deleted on server)

On Desktop:
- polling of notifications

Missing:
iOS @nextcloud/ios:
- [x] "delete" action on push notification is missing
- [x] handle incoming "delete" push notification

Android @nextcloud/android:
- [x] "delete" action on push notification is missing https://github.com/nextcloud/android/pull/3608 https://github.com/nextcloud/android/pull/3606
- [x] "delete" action in notification list is missing  https://github.com/nextcloud/android/pull/3607
- [x] handle incoming "delete" push notification https://github.com/nextcloud/android/pull/3969

On server @nickvergessen:
- [x] once a notification is dismissed this should be sent as "deleted" event with id to all devices so they can remove the pending push notification

Desktop @nextcloud/desktop:
- [ ] notifications should be on top
- [x] notifications should be removed once they are dismissed on server https://github.com/nextcloud/desktop/issues/685

### LINKED ISSUE #13980: Sync (Push) notifications
Done
- [x] Android
- [ ] iOS
- [ ] Desktop
- [x] Server

When receiving a notification:

On Android/iOS:
- you get a push notification (displayed as popup/ in statusbar)
- you can click on it, get redirected to list of notifications
- on iOS you can delete notifications via swipe (get also deleted on server)

On Desktop:
- polling of notifications

Missing:
iOS @nextcloud/ios:
- [ ] "delete" action on push notification is missing
- [ ] handle incoming "delete" push notification

Android @nextcloud/android:
- [x] "delete" action on push notification is missing https://github.com/nextcloud/android/pull/3608 https://github.com/nextcloud/android/pull/3606
- [x] "delete" action in notification list is missing  https://github.com/nextcloud/android/pull/3607
- [x] handle incoming "delete" push notification https://github.com/nextcloud/android/pull/3969

On server @nickvergessen:
- [x] once a notification is dismissed this should be sent as "deleted" event with id to all devices so they can remove the pending push notification

Desktop @nextcloud/desktop:
- [ ] notifications should be on top
- [x] notifications should be removed once they are dismissed on server https://github.com/nextcloud/desktop/issues/685

### PR DESCRIPTION
**Pull request by nickvergessen**
- [x] New "state of the art" registration method for notifiers and apps, this breaks, yes. But it's intentionally to make the devs aware of all the changes. Also I will provide PRs where possible
- [x] Added return types and parameter type hints
- [x] New action type "WEB" to allow to redirect to a given page. E.g. for calls, because you can not start a call on any page (yet)
- [x] Unify the way how notifications are marked "processed" when they should be deleted, also to allow … v
- [x] Silent push notification when one is deleted.
- [x] Notification app adjustments: https://github.com/nextcloud/notifications/pull/318
- [x] Update all shipped notifiers
- [x] Send PRs to all other notifiers

Fix nextcloud/desktop#8697

### CODE REVIEW

**Comment by MorrisJobke** [comment]
@nickvergessen What is the status here? We are close to the beta 1. Should this go into 17 or 18?

**Comment by nickvergessen** [comment]
17 please, I just rebased to fix all the tests and send PRs to all affected apps

**Comment by nickvergessen** [comment]
Ready to review.

Would also be nice if you can check the notifications PR (see PR description) and the PRs to all affected apps (linked  around the last comment from morris)

**Comment by MorrisJobke** [comment]
Missing apps:

* cospend: https://gitlab.com/eneiluj/cospend-nc/blob/master/appinfo/app.php#L21
* gravatar: https://github.com/weeman1337/nextcloud-gravatar/blob/fc078a0023f926262fb65936e92aede1ebf09837/lib/AppInfo/Application.php#L108
* passwords: https://git.mdns.eu/nextcloud/passwords/blob/master/src/lib/AppInfo/Application.php#L207
* passman: https://github.com/nextcloud/passman/blob/f3ba9899b85f41116973d2c4f098df5a92ad0a61/appinfo/app.php#L29
* phonetrack: https://gitlab.com/eneiluj/phonetrack-oc/blob/master/appinfo/app.php#L21
* preferred_providers: https://github.com/nextcloud/preferred_providers/blob/bf9cd16528a3c92a96479e2caef89229ea839cf4/lib/AppInfo/Application.php#L48
* suspicious_login: https://github.com/nextcloud/suspicious_login/blob/ee8acb75343b1616d6d8557a75b913ad4c7349e0/lib/AppInfo/BootstrapSingleton.php#L99
* W2G2: https://github.com/newroco/W2G2/blob/1ae5c6177bfa7e4cb206972f63353d210e449ce0/w2g2/lib/AppInfo/Application.php#L30

File: lib/private/Notification/Manager.php [hunk 1]
```diff
@@ -34,129 +38,131 @@
...
+		$this->notifierClasses = [];
 		$this->preparingPushNotification = false;
 	}
-
 	/**
-	 * @param \Closure $service The service must implement IApp, otherwise a
+	 * @param string $appClass The service must implement IApp, otherwise a
 	 *                          \InvalidArgumentException is thrown later
-	 * @since 8.2.0
+	 * @since 9.0.0
```
Comment (MorrisJobke) [inline lib/private/Notification/Manager.php]: Shouldn't it be `17`, because the parameter has changed?

**Review by MorrisJobke — APPROVED** [review]
Beside my little nitpick the code change looks good 👍

File: lib/private/Notification/Manager.php (hunk 1 above)
Comment (nickvergessen) [reply lib/private/Notification/Manager.php]: yeah, looks like a copy paste error

**Comment by MorrisJobke** [comment]
Tested and works 👍

**Comment by nickvergessen** [comment]
> Missing apps:
>    * cospend: https://gitlab.com/eneiluj/cospend-nc/blob/master/appinfo/app.php#L21
>    * gravatar: https://github.com/weeman1337/nextcloud-gravatar/blob/fc078a0023f926262fb65936e92aede1ebf09837/lib/AppInfo/Application.php#L108
>    * passwords: https://git.mdns.eu/nextcloud/passwords/blob/master/src/lib/AppInfo/Application.php#L207
>    * passman: https://github.com/nextcloud/passman/blob/f3ba9899b85f41116973d2c4f098df5a92ad0a61/appinfo/app.php#L29
>    * phonetrack: https://gitlab.com/eneiluj/phonetrack-oc/blob/master/appinfo/app.php#L21
>    * W2G2: https://github.com/newroco/W2G2/blob/1ae5c6177bfa7e4cb206972f63353d210e449ce0/w2g2/lib/AppInfo/Application.php#L30

Created an issue on all repos and emailed the one where I couldn't due to permissions

File: lib/private/Notification/Manager.php [hunk 2]
```diff
@@ -34,129 +38,131 @@
...
 	 *                          the app id and the app name
+	 * @deprecated 17.0.0 use registerNotifierService instead.
 	 * @since 8.2.0 - Parameter $info was added in 9.0.0
 	 */
 	public function registerNotifier(\Closure $service, \Closure $info) {
-		$this->notifiersClosures[] = $service;
-		$this->notifiersInfoClosures[] = $info;
-		$this->notifiers = [];
-		$this->notifiersInfo = [];
+		$infoData = $info();
```
Comment (rullzer) [inline lib/private/Notification/Manager.php]: Would it make sense to just keep allowing this registration for now (with extra logging). So we don't force all apps to hard update? Like keep it as a compatibiloity layer for 2 releases or so?

Comment (nickvergessen) [reply lib/private/Notification/Manager.php]: As discussed this is not the best idea with the all the changes.
Most obvious the type hints, but also the changes on how deletion is propagated and therefor should be performed correctly when being done in the prepare method of a notifier.

Also all notifiers from the app store have been patched, apart from the 6 which got an issue reported instead due to missing permissions

Comment (juliushaertl) [reply lib/private/Notification/Manager.php]: The issue is that apps can then no longer ship a 16 and 17 compatible version, right?

Comment (nickvergessen) [reply lib/private/Notification/Manager.php]: Yeah, which they can't anyway with the behaviour changes how notifications should be deleted on rendering etc.

Comment (rullzer) [reply lib/private/Notification/Manager.php]: Right. I remember it now. Yeah makes sense then

**Review by blizzz — APPROVED** [review]
changes look good

**Comment by skjnldsv** [comment]
@nickvergessen Our deprecation guidelines is to wait 3 releases before removing the old legacy way. Now on latest master everything is failing for me.

I get empty css on almost all apps that uses the old notifications way because it gets thrown before.

Capture d’écran_2019-07-28_10-16-08


```json
{
  "reqId": "H90SnKlyMmWFtO2itxYU",
  "level": 3,
  "time": "2019-07-28T08:16:57+00:00",
  "remoteAddr": "172.21.0.1",
  "user": "Test1",
  "app": "no app in context",
  "method": "GET",
  "url": "/ocs/v2.php/apps/spreed/api/v1/room",
  "message": {
    "Exception": "InvalidArgumentException",
    "Message": "Notifier Talk (id: spreed) is not considered because it is using the old way to register.",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/html/apps2/spreed/lib/AppInfo/Application.php",
        "line": 136,
        "function": "registerNotifier",
        "class": "OC\\Notification\\Manager",
        "type": "->",
        "args": [
          {
            "__class__": "Closure"
          },
          {
            "__class__": "Closure"
          }
        ]
      },
      {
        "file": "/var/www/html/apps2/spreed/lib/AppInfo/Application.php",
        "line": 99,
        "function": "registerNotifier",
        "class": "OCA\\Spreed\\AppInfo\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "OC\\Server"
          }
        ]
      },
      {
        "file": "/var/www/html/apps2/spreed/appinfo/app.php",
        "line": 24,
        "function": "register",
       {
  "reqId": "H90SnKlyMmWFtO2itxYU",
  "level": 3,
  "time": "2019-07-28T08:16:57+00:00",
  "remoteAddr": "172.21.0.1",
  "user": "Test1",
  "app": "no app in context",
  "method": "GET",
  "url": "/ocs/v2.php/apps/spreed/api/v1/room",
  "message": {
    "Exception": "InvalidArgumentException",
    "Message": "Notifier Talk (id: spreed) is not considered because it is using the old way to register.",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/html/apps2/spreed/lib/AppInfo/Application.php",
        "line": 136,
        "function": "registerNotifier",
        "class": "OC\\Notification\\Manager",
        "type": "->",
        "args": [
          {
            "__class__": "Closure"
          },
          {
            "__class__": "Closure"
          }
        ]
      },
      {
        "file": "/var/www/html/apps2/spreed/lib/AppInfo/Application.php",
        "line": 99,
        "function": "registerNotifier",
        "class": "OCA\\Spreed\\AppInfo\\Application",
        "type": "->",
        "args": [
          {
            "__class__": "OC\\Server"
          }
        ]
      },
      {
        "file": "/var/www/html/apps2/spreed/appinfo/app.php",
        "line": 24,
        "function": "register",
        "class": "OCA\\Spreed\\AppInfo\\Application",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/html/lib/private/legacy/app.php",
        "line": 262,
        "args": [
          "/var/www/html/apps2/spreed/appinfo/app.php"
        ],
        "function": "require_once"
      },
      {
        "file": "/var/www/html/lib/private/legacy/app.php",
        "line": 155,
        "function": "requireAppFile",
        "class": "OC_App",
        "type": "::",
        "args": [
          {
            "__class__": "OCA\\Spreed\\AppInfo\\Application"
          }
        ]
      },
      {
        "file": "/var/www/html/lib/private/legacy/app.php",
        "line": 128,
        "function": "loadApp",
        "class": "OC_App",
        "type": "::",
        "args": [
          "spreed"
        ]
      },
      {
        "file": "/var/www/html/ocs/v1.php",
        "line": 56,
        "function": "loadApps",
        "class": "OC_App",
        "type": "::",
        "args": []
      },
      {
        "file": "/var/www/html/ocs/v2.php",
        "line": 24,
        "args": [
          "/var/www/html/ocs/v1.php"
        ],
        "function": "require_once"
      }
    ],
    "File": "/var/www/html/lib/private/Notification/Manager.php",
    "Line": 86,
    "CustomMessage": "--"
  },
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36",
  "version": "17.0.0.1"
} "class": "OCA\\Spreed\\AppInfo\\Application",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/html/lib/private/legacy/app.php",
        "line": 262,
        "args": [
          "/var/www/html/apps2/spreed/appinfo/app.php"
        ],
        "function": "require_once"
      },
      {
        "file": "/var/www/html/lib/private/legacy/app.php",
        "line": 155,
        "function": "requireAppFile",
        "class": "OC_App",
        "type": "::",
        "args": [
          {
            "__class__": "OCA\\Spreed\\AppInfo\\Application"
          }
        ]
      },
      {
        "file": "/var/www/html/lib/private/legacy/app.php",
        "line": 128,
        "function": "loadApp",
        "class": "OC_App",
        "type": "::",
        "args": [
          "spreed"
        ]
      },
      {
        "file": "/var/www/html/ocs/v1.php",
        "line": 56,
        "function": "loadApps",
        "class": "OC_App",
        "type": "::",
        "args": []
      },
      {
        "file": "/var/www/html/ocs/v2.php",
        "line": 24,
        "args": [
          "/var/www/html/ocs/v1.php"
        ],
        "function": "require_once"
      }
    ],
    "File": "/var/www/html/lib/private/Notification/Manager.php",
    "Line": 86,
    "CustomMessage": "--"
  },
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36",
  "version": "17.0.0.1"
}
```

**Comment by nickvergessen** [comment]
This has been discussed in length above and in the linked issues.

**Comment by nickvergessen** [comment]
But then again, the exception is only logged, not thrown and it should therefor also not fail somewhere resulting in such pages.

**Comment by skjnldsv** [comment]
@nickvergessen yes, you're right!
After upgrading spreed, it worked again (though, all my instance was broken because of this)

Nonetheless, I think we should respect the deprecation process we have as much as we can :)
But maybe we couldn't avoid this! I was just surprised ;)

Nice work though! Much cleaner code!

**Comment by MorrisJobke** [comment]
> @nickvergessen Our deprecation guidelines is to wait 3 releases before removing the old legacy way. 

@skjnldsv It's years and not releases. We are waiting 3 years until deprecated stuff is removed.

**Pull request merged**